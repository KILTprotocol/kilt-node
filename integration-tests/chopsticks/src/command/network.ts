import { mainChains } from '../network/index.js'
import { setStorage, setupNetwork } from '../network/utils.js'

export async function createTestNetwork() {
	const relayConfig = mainChains.polkadot.getConfig({})

	// 1. Create the network specific state
	const kiltStorage = mainChains.kilt.storage.assignNativeTokensToAccounts([
		'4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr',
	])

	// 2. Create the network specific config
	const kiltConfig = mainChains.kilt.getConfig({})

	// 2. Add the different parachainContexts here. The order of parachainContexts and parachainStorage should match.
	const parachainOptions = [kiltConfig]
	const parachainStorage = [kiltStorage]

	const { relayChainContext, parachainContexts } = await setupNetwork(relayConfig, parachainOptions)

	await Promise.all(parachainStorage.map((storage, index) => setStorage(parachainContexts[index], storage)))

	await Promise.all([...parachainContexts, relayChainContext].map((context) => context.pause()))
}
