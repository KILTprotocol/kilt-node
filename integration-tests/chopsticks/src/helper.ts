import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

console.log('alice', u8aToHex(decodeAddress('4qPZ8fv6BjGoGKzfx5LtBFnEUp2b5Q5C1ErrjBNGmoFTLNHG')))
console.log('alice 2', u8aToHex(decodeAddress(keysAlice.address)))
