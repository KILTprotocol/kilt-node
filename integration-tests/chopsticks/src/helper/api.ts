/* eslint-disable @typescript-eslint/no-explicit-any */
import { ApiPromise } from '@polkadot/api'

export const xtokens = {
	parachainV2: (paraId: number) => (acc: any) => ({
		V1: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: paraId },
					{
						AccountId32: {
							network: 'Any',
							id: acc,
						},
					},
				],
			},
		},
	}),
	parachainV3: (paraId: number) => (acc: any) => ({
		V3: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: paraId },
					{
						AccountId32: {
							id: acc,
						},
					},
				],
			},
		},
	}),
	transfer:
		(token: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.xTokens.transfer(token, amount, dest(acc), weight),
}

export const balances = {
	transferAllowDeath: ({ api }: { api: ApiPromise }, dest: string, amount: string) =>
		api.tx.balances.transferAllowDeath(dest, amount),
}

export const switchPallet = {
	switchV3:
		() =>
		({ api }: { api: ApiPromise }, acc: any, amount: string) =>
			api.tx.assetSwitchPool1.switch(amount, {
				V3: {
					parents: 0,
					interior: {
						X1: {
							AccountId32: {
								id: acc,
							},
						},
					},
				},
			}),
	switchV4:
		() =>
		({ api }: { api: ApiPromise }, acc: any, amount: string) =>
			api.tx.assetSwitchPool1.switch(amount, {
				V4: {
					parents: 0,
					interior: {
						X1: [
							{
								AccountId32: {
									id: acc,
								},
							},
						],
					},
				},
			}),
	resume:
		() =>
		({ api }: { api: ApiPromise }) =>
			api.tx.assetSwitchPool1.resumeSwitchPair(),
	pause:
		() =>
		({ api }: { api: ApiPromise }) =>
			api.tx.assetSwitchPool1.pauseSwitchPair(),
}

export const xcmPallet = {
	parachainV2: (parents: number, paraId: number) => ({
		V2: {
			parents,
			interior: {
				X1: { Parachain: paraId },
			},
		},
	}),
	parachainV3: (parents: number, paraId: any) => ({
		V3: {
			parents,
			interior: {
				X1: { Parachain: paraId },
			},
		},
	}),
	parachainV4: (parents: number, paraId: any) => ({
		V4: {
			parents,
			interior: { X1: [{ Parachain: paraId }] },
		},
	}),
	limitedTeleportAssets:
		(token: any, amount: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedTeleportAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV2:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V2: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									network: 'Any',
									id: acc,
								},
							},
						},
					},
				},
				{
					V2: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV3:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV4:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V4: {
						parents: 0,
						interior: {
							X1: [
								{
									AccountId32: {
										id: acc,
									},
								},
							],
						},
					},
				},
				{
					V4: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsV3:
		(dest: any, token: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.polkadotXcm.transferAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsV4:
		(dest: any, token: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.polkadotXcm.transferAssets(
				dest,
				{
					V4: {
						parents: 0,
						interior: {
							X1: [
								{
									AccountId32: {
										id: acc,
									},
								},
							],
						},
					},
				},
				{
					V4: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsUsingTypeAndThenV4:
		(dest: any, token: any, xcmMessage: any) =>
		({ api }: { api: ApiPromise }, balanceToTransfer: string) =>
			api.tx.polkadotXcm.transferAssetsUsingTypeAndThen(
				dest,
				{
					V4: [
						{
							id: token,
							fun: { Fungible: balanceToTransfer },
						},
					],
				},
				'LocalReserve',
				{ V4: token },
				'LocalReserve',
				xcmMessage,
				'Unlimited'
			),
	transferAssetsUsingTypeAndThenV3:
		(dest: any, token: any, xcmMessage: any) =>
		({ api }: { api: ApiPromise }, balanceToTransfer: string) =>
			api.tx.polkadotXcm.transferAssetsUsingTypeAndThen(
				dest,
				{
					V3: [
						{
							id: token,
							fun: { Fungible: balanceToTransfer },
						},
					],
				},
				'LocalReserve',
				{ V3: token },
				'LocalReserve',
				xcmMessage,
				'Unlimited'
			),
	send:
		(destination: object) =>
		({ api }: { api: ApiPromise }, xcmMessage: object) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).send(destination, xcmMessage),
}

/**
 * Different pallets to submit tx
 */
export const tx = {
	xtokens,
	xcmPallet,
	switchPallet,
	balances,
}

/**
 * Query functions for different chains.
 * Native tokens are fetched via the system pallet, while other tokens are fetched via the tokens or assets pallet.
 *
 */
export const query = {
	balances: async ({ api }: { api: ApiPromise }, address: string) =>
		((await api.query.system.account(address)) as any).data.free.toBigInt(),
	foreignAssets:
		(assetId: any) =>
		async ({ api }: { api: ApiPromise }, address: string) => {
			const accountInfo: any = await api.query.foreignAssets.account(assetId, address)
			if (accountInfo.isNone) {
				return 0n
			}
			return accountInfo.unwrap().balance.toBigInt()
		},
	fungibles:
		(assetId: any) =>
		async ({ api }: { api: ApiPromise }, address: string) => {
			const accountInfo: any = await api.query.fungibles.account(assetId, address)
			if (accountInfo.isNone) {
				return 0n
			}
			return accountInfo.unwrap().balance.toBigInt()
		},
	tokens:
		(token: any) =>
		async ({ api }: { api: ApiPromise }, address: string) =>
			((await api.query.tokens.accounts(address, token)) as any).data.free.toBigInt(),
}
