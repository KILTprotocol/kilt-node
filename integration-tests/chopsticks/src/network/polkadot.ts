import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceDOT, toNumber } from '../utils.js'

/// Options used to create the HydraDx context
export const options: SetupOption = {
	endpoint: process.env.POLKADOT_WS || [
		'wss://rpc.polkadot.io',
		'wss://polkadot-rpc.dwellir.com',
		'wss://rpc.ibp.network/polkadot',
	],
	db: './db/polkadot.db.sqlite',
	port: toNumber(process.env.POLKADOT_PORT) || 9000,
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceDOT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

export function removeDisputesAndMessageQueues() {
	return {
		ParasDisputes: {
			// those can makes block building super slow
			$removePrefix: ['disputes'],
		},
		Dmp: {
			// clear existing dmp to avoid impact test result
			$removePrefix: ['downwardMessageQueues'],
		},
	}
}

export async function getContext(): Promise<Config> {
	return setupContext(options)
}
