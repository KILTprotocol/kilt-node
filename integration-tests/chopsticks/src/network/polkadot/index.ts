import * as Main from './main_network.js'
import * as Test from './test_network.js'

export default {
	main: {
		storage: Main.storage,
		getSetupOptions: Main.getSetupOptions,
	},
	test: {
		storage: Test.storage,
		getSetupOptions: Test.getSetupOptions,
	},
}
