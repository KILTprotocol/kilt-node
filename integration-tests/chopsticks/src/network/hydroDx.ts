import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'
import * as SpiritnetConfig from './spiritnet.js'

export const options: SetupOption = {
	endpoint: ['wss://hydradx-rpc.dwellir.com', 'wss://rpc.hydradx.cloud'],
	db: './db/hydradx.db.sqlite',
	port: 8001,
}

const kiltTokenId = 60

export const defaultStorage = (addr: string) => ({
	TechnicalCommittee: { Members: [addr] },
	Council: { Members: [addr] },
	Tokens: {
		Accounts: [[[addr, kiltTokenId], { free: 100 * 10e12 }]],
	},
	assetRegistry: {
		assetLocations: [[[kiltTokenId], { parents: 1, interior: { X1: { Parachain: SpiritnetConfig.paraId } } }]],
		assetIds: [[['KILT'], kiltTokenId]],
		locationAssets: [[[{ parents: 1, interior: { X1: { Parachain: SpiritnetConfig.paraId } } }], kiltTokenId]],
		assets: [
			[
				[kiltTokenId],
				{
					name: 'KILT',
					assetType: 'Token',
					existentialDeposit: 500,
					symbol: 'KLT',
					decimals: 18,
					xcmRateLimit: null,
					isSufficient: true,
				},
			],
		],
	},
	multiTransactionPayment: {
		acceptedCurrencies: [[[kiltTokenId], 100_000]],
	},

	System: {
		Account: [[[addr], { providers: 1, data: { free: 100 * 10e12 } }]],
	},
})

export const paraId = 2034
export const sovereignAccount = u8aToHex(decodeAddress('5Eg2fntQqFi3EvFWAf71G66Ecjjah26bmFzoANAeHFgj9Lia'))

export async function getContext(): Promise<Config> {
	return setupContext(options)
}
