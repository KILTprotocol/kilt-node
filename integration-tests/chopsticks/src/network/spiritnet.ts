import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types'
import * as HydraDxConfig from './hydroDx'

export const options: SetupOption = {
	endpoint: 'wss://kilt-rpc.dwellir.com',
	db: './db/spiritnet.db.sqlite',
	port: 8002,
	wasmOverride: '../../target/debug/wbuild/spiritnet-runtime/spiritnet_runtime.wasm',
	allowUnresolvedImports: true,
	timeout: 600000,
	runtimeLogLevel: 5,
}

export const defaultStorage = (addr: string) => ({
	// set technical committee and council to addr
	technicalCommittee: { Members: [addr] },
	council: { Members: [addr] },
	System: {
		Account: [[[addr], { providers: 1, data: { free: 1000 * 1e12 } }]],
	},
	polkadotXcm: {
		safeXcmVersion: 3,
	},
})

export const spiritnet = {
	paraId: 2086,
	hydraDxDestination: {
		V2: {
			parents: 1,
			interior: {
				X1: {
					Parachain: HydraDxConfig.paraId,
				},
			},
		},
	},
	hydraDxBeneficiary: {
		V2: {
			parents: 1,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: HydraDxConfig.sovereignAccount,
					},
				},
			},
		},
	},
}

export async function getContext(): Promise<Config> {
	return setupContext(options)
}
