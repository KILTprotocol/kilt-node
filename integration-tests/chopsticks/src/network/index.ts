import * as SpiritnetConfig from './spiritnet.js'
import * as PolkadotConfig from './relay.js'
import * as HydraDxConfig from './hydraDx.js'
import { ChainConfigs } from './types.js'

/*
 * Get an environment variable and throw an error if it is not set.
 */
function getEnvVariable(name: string): string {
	const value = process.env[name]
	if (value === undefined) {
		throw new Error(`Environment variable ${name} is not set.`)
	}
	return value
}

/*
 * Object with all the chain configurations.
 */
export const all: ChainConfigs = {
	spiritnet: {
		getConfig: SpiritnetConfig.getSetupOptions,
		parameters: {
			blockNumber: Number(getEnvVariable('SPIRITNET_BLOCK_NUMBER')),
			wasmOverride: getEnvVariable('SPIRITNET_WASM_OVERRIDE'),
		},
	},
	hydraDx: {
		getConfig: HydraDxConfig.getSetupOptions,
		parameters: {
			blockNumber: Number(getEnvVariable('HYDRADX_BLOCK_NUMBER')),
		},
	},
	polkadot: {
		getConfig: PolkadotConfig.getSetupOptions,
		parameters: {
			blockNumber: Number(getEnvVariable('POLKADOT_BLOCK_NUMBER')),
		},
	},
}
