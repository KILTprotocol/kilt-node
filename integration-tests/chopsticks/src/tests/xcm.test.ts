import { test, beforeAll, afterAll } from 'vitest'
import { connectParachains, connectVertical, xcmLogger } from '@acala-network/chopsticks'
import { sendTransaction } from '@acala-network/chopsticks-testing'

import * as SpiritnetNetwork from '../network/spiritnet'
import * as PolkadotNetwork from '../network/polkadot'
import * as HydraDxNetwork from '../network/hydroDx'
import type { Config } from '../network/types'
import { checkEvents, keysBob, keysCharlie } from '../helper'

let spiritnetContext: Config
let hydradxContext: Config
let polkadotContext: Config

beforeAll(async () => {
	xcmLogger.level = 'info'
	spiritnetContext = await SpiritnetNetwork.getContext()
	hydradxContext = await HydraDxNetwork.getContext()
	polkadotContext = await PolkadotNetwork.getContext()

	await polkadotContext.dev.setStorage(PolkadotNetwork.defaultStorage(keysCharlie.address))
	await spiritnetContext.dev.setStorage(SpiritnetNetwork.defaultStorage(keysBob.address))

	// Setup network
	await connectVertical(polkadotContext.chain, spiritnetContext.chain)
	await connectVertical(polkadotContext.chain, hydradxContext.chain)
	await connectParachains([spiritnetContext.chain, hydradxContext.chain])

	await polkadotContext.dev.newBlock()
	console.log('polkadot created block')
	await spiritnetContext.dev.newBlock()
	console.log('Spiritnet created block')
	await hydradxContext.dev.newBlock()

	await polkadotContext.dev.newBlock()
	console.log('polkadot created block')
	await spiritnetContext.dev.newBlock()
	console.log('Spiritnet created block')
	await hydradxContext.dev.newBlock()

	await new Promise((res) => {
		setTimeout(res, 10000)
	})
	console.log('Network is created')

	// Perform runtime upgrade

	console.log('Runtime Upgrade completed')
}, 40_000)

afterAll(() => {
	// spiritnetContext.teardown()
	// hydradxContext.teardown()
	// polkadotContext.teardown()
})

test(
	'Limited Reserve Transfers from Spiritnet Account Bob -> HydraDx',
	async () => {
		const signedTx = spiritnetContext.api.tx.polkadotXcm
			.limitedReserveTransferAssets(
				SpiritnetNetwork.spiritnet.hydraDxDestination,
				SpiritnetNetwork.spiritnet.hydraDxBeneficiary,
				{
					V2: [
						{
							id: { Concrete: { parents: 0, interior: 'Here' } },
							fun: { Fungible: 1 * 10e12 },
						},
					],
				},
				0,
				'Unlimited'
			)
			.signAsync(keysBob)

		const events = await sendTransaction(signedTx)

		await spiritnetContext.chain.newBlock()

		checkEvents(events, 'balances').toMatchSnapshot('Balance events')
		checkEvents(events, 'polkadotXcm').toMatchSnapshot('sender events')

		while (1) {}
	},
	{ timeout: 100_000 }
)
