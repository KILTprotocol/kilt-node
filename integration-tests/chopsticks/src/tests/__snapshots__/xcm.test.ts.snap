// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Limited Reserve Transfers from Spiritnet Account Bob -> HydraDx > Balance events 1`] = `
[
  {
    "data": {
      "who": "4qpE21nvgo8AmyNMi32T7r4LWitN7fJaUox4PmGRUJvGqH7W",
    },
    "method": "Upgraded",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 170000000000)",
      "who": "4qpE21nvgo8AmyNMi32T7r4LWitN7fJaUox4PmGRUJvGqH7W",
    },
    "method": "Withdraw",
    "section": "balances",
  },
  {
    "data": {
      "account": "4qXPdpioJ6D8cgdeYXaukV2Y2oAQUHaX1VnGhdbSRqJn2CBt",
      "freeBalance": 10000000000000,
    },
    "method": "Endowed",
    "section": "balances",
  },
  {
    "data": {
      "amount": 10000000000000,
      "from": "4qpE21nvgo8AmyNMi32T7r4LWitN7fJaUox4PmGRUJvGqH7W",
      "to": "4qXPdpioJ6D8cgdeYXaukV2Y2oAQUHaX1VnGhdbSRqJn2CBt",
    },
    "method": "Transfer",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 84000000000)",
      "who": "4qPZ8fv6BjGoGKzfx5LtBFnEUp2b5Q5C1ErrjBNGmoFTLNHG",
    },
    "method": "Deposit",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 84000000000)",
      "who": "4oEapGbe4JYognwWu5jN4MZsLAtbWAfVwo2gqppxqudtS6jx",
    },
    "method": "Deposit",
    "section": "balances",
  },
]
`;

exports[`Limited Reserve Transfers from Spiritnet Account Bob -> HydraDx > sender events 1`] = `
[
  {
    "data": {
      "outcome": {
        "Incomplete": [
          {
            "proofSize": 0,
            "refTime": 400000000,
          },
          "Transport",
        ],
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;
